src/include/migraphx/operators.hpp

```diff
@@ -810,41 +810,210 @@ struct gather
+            auto a_lens = inputs[0].lens();
+            auto b_lens = inputs[1].lens();
+            check_dims(a_lens, b_lens);
+
+            auto out_lens   = a_lens;
+            auto t          = inputs[0].type();
+            out_lens.back() = b_lens.back();
+
+            // C should be the same shape as A * B
+            auto c_lens = inputs[2].lens();
+            if(c_lens != out_lens)
+            {
+                MIGRAPHX_THROW("DOT: dimension mismatch, operand C: {" + to_string_range(c_lens) +
+                               "}, cannot add to operand A * B: {" + to_string_range(out_lens) +
+                               "}");
+            }
+
+            return {t, out_lens};
+        }
+
+        // For the case of two inputs, it is the numpy.matmul
         check_shapes{inputs, *this}.has(2).same_type();
         const shape& a = inputs.at(0);
         const shape& b = inputs.at(1);
         auto t         = a.type();
 
-        // according to the specification of the numpy.matmul()
-        // inputs with the shape dims more than 2 are acceptable
-        // as long as dim values are the same in the two inputs
-        if(!std::equal(a.lens().rbegin() + 2, a.lens().rend(), b.lens().rbegin() + 2))
+        if(a.scalar() || b.scalar())
         {
-            MIGRAPHX_THROW("DOT: dim values mismatch");
+            MIGRAPHX_THROW("DOT: scalar operands are not allowed, use op::mul{} instead");
         }
 
-        std::size_t dim_0 = a.lens().size() - 2;
-        std::size_t dim_1 = a.lens().size() - 1;
-        if(a.lens()[dim_1] != b.lens()[dim_0])
-            MIGRAPHX_THROW("Inner dimensions do not match: {" + to_string_range(a.lens()) +
-                           "} x {" + to_string_range(b.lens()) + "}");
-        auto out_lens   = a.lens();
-        out_lens[dim_1] = b.lens()[dim_1];
-        return {t, out_lens};
+        auto a_lens        = a.lens();
+        auto b_lens        = b.lens();
+        bool is_a_appended = false;
+        bool is_b_appended = false;
+
+        if(a_lens.size() == 1)
+        {
+            a_lens.insert(a_lens.begin(), 1);
+            is_a_appended = true;
+        }
+
+        if(b_lens.size() == 1)
+        {
+            b_lens.push_back(1);
+            is_b_appended = true;
+        }
+
+        std::size_t dim_1 = a_lens.size() - 1;
+        std::size_t dim_0 = b_lens.size() - 2;
+        if(a_lens[dim_1] != b_lens[dim_0])
+        {
+            MIGRAPHX_THROW("DOT : dimension mismatch, operand A: {" + to_string_range(a.lens()) +
+                           "}, cannot multiply operand B: {" + to_string_range(b.lens()) + "}");
+        }
+
+        // remove the matrix dims, do multi_broadcast of the shape of the batch
+        a_lens.pop_back();
+        std::size_t out_m = a_lens.back();
+        a_lens.pop_back();
+
+        std::size_t out_n = b_lens.back();
+        b_lens.pop_back();
+        b_lens.pop_back();
+
+        auto out_lens = shape_broadcast(a_lens, b_lens);
```

Comment:
Why is this still necessary if the user(or onnx parser) is adding the multibroadcast operators?